src/map/achievement.c

```diff
@@ -0,0 +1,1693 @@
+
+	/* Status Types */
+	equal = (objective->unique.status_type && objective->unique.status_type == criteria->unique.status_type);
+
+	/* Item Type */
+	equal = (objective->item_type && (objective->item_type&criteria->item_type) != 0);
+
+	/* Job Ids */
+	for (i = 0; i < VECTOR_LENGTH(objective->jobid); i++) {
+		ARR_FIND(0, VECTOR_LENGTH(criteria->jobid), j, VECTOR_INDEX(criteria->jobid, j) == VECTOR_INDEX(objective->jobid, i));
+		equal = (j < VECTOR_LENGTH(criteria->jobid));
+	}
+
+	return equal;
+}
+
+/**
+ * Validates an Achievement Objective of similar types.
+ * @param sd          as a pointer to the map session data.
+ * @param type        as the type of the achievement.
+ * @param criteria    as the criteria of the objective (mob id, job id etc.. 0 for no criteria).
+ * @param progress    as the current progress of the objective.
+ * @return total number of updated achievements on success, 0 on failure.
+ */
+int achievement_validate_type(struct map_session_data *sd, const enum achievement_types type, const struct achievement_objective *criteria, const int progress, bool additive)
+{
+	int i = 0, total = 0;
+	struct achievement *ach = NULL;
+
+	nullpo_ret(sd);
+	Assert_ret(progress > 0);
+
+	if (type == ACH_QUEST) {
+		ShowDebug("achievement_validate_type: ACH_QUEST is not handled by this function. (use achievement_validate())\n");
+		return 0;
+	} else if (type >= ACH_TYPE_MAX) {
+		ShowDebug("achievement_validate_type: Invalid Achievement Type %d! (min: %d, max: %d)\n", (int)type, (int)ACH_QUEST, (int)ACH_TYPE_MAX - 1);
+		return 0;
+	}
+
+	/* Loop through all achievements of the type, checking for possible matches. */
+	for (i = 0; i < VECTOR_LENGTH(achievement->category[type]); i++) {
+		int j = 0;
+		bool updated = false;
+		struct achievement_data *ad = NULL;
+
+		if ((ad = achievement->get(VECTOR_INDEX(achievement->category[type], i))) == NULL)
+			continue;
+		
+		for (j = 0; j < VECTOR_LENGTH(ad->objective); j++) {
+			// Check if objective criteria matches.
+			if (achievement->check_criteria(&VECTOR_INDEX(ad->objective, j), criteria) == false)
+				continue;
+			// Criteria passed, check if not completed and update progress.
+			if ((ach = achievement->find_or_create(sd, ad->id, false)) == NULL
+				|| (!ach->completed_at && ach->objective[j] < VECTOR_INDEX(ad->objective, j).goal)) {
+				if (additive)
+					achievement->progress_add(sd, ad, j, progress);
+				else
+					achievement->progress_set(sd, ad, j, progress);
+				updated = true;
+			}
+		}
+
+		if (updated) total++;
+	}
+
+	return total;
+}
+
+/**
+ * Validates any achievement's specific objective index.
+ * @param sd       pointer to the session data.
+ * @param aid      ID of the achievement.
+ * @param index    index of the objective.
+ * @param progress progress to be added towards the goal.
+ */
+bool achievement_validate(struct map_session_data *sd, const int aid, const unsigned int index, const int progress, bool additive)
```

Comment:
Not really, I actually did not think of HPM. Simply because they weren't meant to be changed within the function but allowing so through plugins makes sense.